### PR TITLE
Working

### PR DESCRIPTION
**Pull request by gengem777**
My Ist pull request

### CODE REVIEW

File: pricers.py [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+       u_hat: parameter related to input functions with shape [B, M, N, k], k is the number of parameters
+    """
+    def __init__(self, sde, option, config):
+        super(BaseBSDEPricer, self).__init__()
+        self.eqn_config = config.eqn_config
+        self.net_config = config.net_config
+        self.option = option
+        self.sde = sde
+        self.dim = self.eqn_config.dim # dimension of assets
+        self.branch_layers = self.net_config.branch_layers # a list: neurons of each layer in branch network
```
Comment (owenyoung75) [inline pricers.py]: remember to change the var names to be more accurate ones.

File: pricers.py [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+                                                    trunk_layer=self.trunk_layers, 
+                                                    pi_layer=self.pi_layers, 
+                                                    num_assets=self.dim, 
+                                                    dense=False, 
+                                                    num_outputs=6,
+                                                    activation=self.activation,
+                                                    filters=self.filters, 
+                                                    strides=self.strides) # DeepONet with kernel operator be CNN
+        else:
+            if  self.net_config.kernel_type == "no":
```
Comment (owenyoung75) [inline pricers.py]: In the docstring to be added for configs, explain possible choices for this kernel type and the meanings.

Also, in the code, add comments whenever we have this kind of if-else condition, to explain to reviewers what we wanna achieve.

File: pricers.py [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+                                                    dense=True, 
+                                                    num_outputs=6,
+                                                    activation=self.activation) # DeepONet with kernel operator be dense operator
+            else:                                 
+                self.no_net = DeepKernelONetwithPI(branch_layer=self.branch_layers, 
+                                                    trunk_layer=self.trunk_layers, 
+                                                    pi_layer=self.pi_layers, 
+                                                    num_assets=self.dim, 
+                                                    dense=False, 
+                                                    num_outputs=6,
```
Comment (owenyoung75) [inline pricers.py]: Magic number here -- consider to change it to something derived or inferred from configs?

File: pricers.py [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+        else:
+            if  self.net_config.kernel_type == "no":
+                self.no_net = DeepONet(branch_layer=self.branch_layers, 
+                                       trunk_layer=self.trunk_layers,
+                                       activation = self.activation) # DeepONet without kernel operator
+            else:
+                self.no_net = DeepKernelONetwithoutPI(branch_layer=self.branch_layers, 
+                                                        trunk_layer=self.trunk_layers, 
+                                                        dense=True, 
+                                                        num_outputs=6,
```
Comment (owenyoung75) [inline pricers.py]: Same here, magic number.

File: pricers.py [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+                                                        filters=self.filters, 
+                                                        strides=self.strides) # DeepONet without pi layers
+        self.time_horizon = self.eqn_config.T # the time horizon $T$ of the problem which is fixed
+        self.batch_size = self.eqn_config.batch_size # the batch size of input functions
+        self.samples = self.eqn_config.sample_size # the number of paths sampled on each input functions set
+        self.dt = self.eqn_config.dt # time step size $\Delta t$
+        self.time_steps = self.eqn_config.time_steps # number of time steps $N = T/\Delta t$
+        time_stamp = tf.range(0, self.time_horizon, self.dt)
+        time_stamp = tf.reshape(time_stamp, [1, 1, self.time_steps, 1])
+        self.time_stamp = tf.tile(time_stamp, [self.batch_size, self.samples, 1, 1]) # time tensor of the problem
```
Comment (owenyoung75) [inline pricers.py]: squeeze these 3 lines into one would be better.

File: pricers.py [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+
+    def call(self, inputs: Tuple[tf.Tensor]) -> tf.Tensor:
+        r"""
+        The forward process is discussed under two different circumstance：
+        when the no_net is the instance of DeepONet, then we just need to input the three tuple (t, x, u):
+            t: time tensor with shape [B, M, N, 1]
+            x: state tensor with shape [B, M, N, d], [B, M, N, 2d] under SV model, [B, M, N, 3d] under SV model for path dependent product
+            u: parameter related to input functions with shape [B, M, N, k], k is the number of parameters
+            return: the out put function value with shape [B, M, N, 1]
+        when the no_net is not the instance of DeepONet, then the function values and kernel operaton will be made with the method split_uhat()
```
Comment (owenyoung75) [inline pricers.py]: Great docstring here! So basically you are trying to explain that the inputs would be either 3-tuple or 4-tuple depending on the network type -- learn to elaborate it more explicitly.

PS: typo "operaton"

File: pricers.py [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+            t: time tensor with shape [B, M, N, 1]
+            x: state tensor with shape [B, M, N, d], [B, M, N, 2d] under SV model, [B, M, N, 3d] under SV model for path dependent product
+            u_c: function embedding tensor [B, M, N, k_1]
+            u_p: other parameters tensor [B, M, N, k_2]
+            return: the out put function value with shape [B, M, N, 1]
+        """
+        t, x, u_hat = inputs
+        if type(self.no_net) == DeepONet:
+            y = self.no_net((t, x, u_hat))
+        # print(t.shape, x.shape, u.shape)
```
Comment (owenyoung75) [inline pricers.py]: Usually e shouldn't leave commented code there, unless it's some important alternative functions temporarily undecided to keep or not.

File: pricers.py [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+        else:
+            u_c, u_p = self.sde.split_uhat(u_hat)
+            # print(u_c.shape, u_p.shape)
```
Comment (owenyoung75) [inline pricers.py]: also here.

File: pricers.py [hunk 9]
```diff
@@ -0,0 +1,321 @@
...
+    def payoff_func(self, t: tf.Tensor, x: tf.Tensor, u_hat: tf.Tensor) -> tf.Tensor:
+        """
+        The payoff function of the time horizon and paths based on the given time tensor and state tensor
+        """
+        raise NotImplementedError
+
+
+class MarkovianPricer(BaseBSDEPricer):
+    """
+    This is the class to construct the tain step of the BSDE loss function, the data is generated from three
```
Comment (owenyoung75) [inline pricers.py:235]: Remember to update the class string, as it represents a pricer object now.

File: pricers.py [hunk 10]
```diff
@@ -0,0 +1,321 @@
...
+    external classes: the sde class which yield the data of input parameters. the option class which yields the
+    input function data and give the payoff function and exact option price.
+    """
+
+    def __init__(self, sde, option, config):
+        super(MarkovianPricer, self).__init__(sde, option, config)
+
+    def loss_fn(self, data: Tuple[tf.Tensor], training=None) -> Tuple[tf.Tensor]:
```
Comment (owenyoung75) [inline pricers.py]: This loss functions is heavy in math, so add a docstring to explain what the loss contains and how it is computed.

File: pricers.py [hunk 11]
```diff
@@ -0,0 +1,321 @@
...
+        """
+        :param t: time B x M x (T-1) x 1
+        :param x: state B x M x (T-1) x D
+        :param  u: B x K ->(repeat) B x M x (T-1) x K
+        :return: value (B, M, (T-1), 1), gradient_x (B, M, (T-1), D)
+        """
+        t, x, dw, u_hat = data
+        steps = self.eqn_config.time_steps
+        loss_interior = 0.0
+        u_now = u_hat[:, :, :-1, :]
```
Comment (owenyoung75) [inline pricers.py]: Name convention: use more accurate (longer) name for these intermediate variables whenever the math part is heavy.

File: pricers.py [hunk 12]
```diff
@@ -0,0 +1,321 @@
...
+        x_now = x[:, :, :-1, :]
+        t_now = t[:, :, :-1, :]
+        with tf.GradientTape(watch_accessed_variables=False) as tape:
+            tape.watch(x)
+            f = self.call((t, x, u_hat))
```
Comment (owenyoung75) [inline pricers.py]: I guess you used replace command to update the function name -- call here should just be _self()_

File: pricers.py [hunk 13]
```diff
@@ -0,0 +1,321 @@
...
+        :param  u: B x K ->(repeat) B x M x (T-1) x K
+        :return: value (B, M, (T-1), 1), gradient_x (B, M, (T-1), D)
+        """
+        t, x, dw, u_hat = data
+        steps = self.eqn_config.time_steps
+        loss_interior = 0.0
+        u_now = u_hat[:, :, :-1, :]
+        x_now = x[:, :, :-1, :]
+        t_now = t[:, :, :-1, :]
+        with tf.GradientTape(watch_accessed_variables=False) as tape:
```
Comment (owenyoung75) [inline pricers.py]: consider to group this nabla computation as a group function, which would make the loss calculation much cleaner, and also it produces a meaningful quantity.

File: pricers.py [hunk 14]
```diff
@@ -0,0 +1,321 @@
...
+            tape.watch(x)
+            f = self.call((t, x, u_hat))
+            grad = tape.gradient(f, x) 
+        f_now = f[:, :, :-1, :]
+        f_pls = f[:, :, 1:, :]
+        grad = grad[:,:,:-1,:]
+        for n in range(steps-1):
```
Comment (owenyoung75) [inline pricers.py]: Actually, we could group interior loss as a separate function as well.

File: pricers.py [hunk 15]
```diff
@@ -0,0 +1,321 @@
...
+        return self.sde.driver_bsde(t, x, y, u_hat) 
+
+    def diffusion_bsde(self, t: tf.Tensor, x: tf.Tensor, grad: tf.Tensor, dw: tf.Tensor, u_hat: tf.Tensor) -> tf.Tensor:
+        """
+        grad: (B, M, N-1, d)
+        dw: (B, M, N-1, d)
+        give: \sigma(t, x) * grad
+        for a batch of (t, x, par)
+        """
+        if not isinstance(self.sde, HestonModel):
```
Comment (owenyoung75) [inline pricers.py]: Add comments here for if-else conditions.

File: pricers.py [hunk 16]
```diff
@@ -0,0 +1,373 @@
...
+        """
+        payoff = self.option.payoff(t, x, u_hat)
+        return payoff
+
+class EuropeanSolver(MarkovianPricer):
+    def __init__(self, sde, option, config):
+        super(EuropeanSolver, self).__init__(sde, option, config)
+        self.num_of_time_intervals_for_early_exercise = 0 # this gives the index of the round of training
+        if self.net_config.pi == "true":
+            if self.net_config.kernel_type == "dense":
```
Comment (owenyoung75) [inline pricers.py:382]: Again, we need to specify all candidates for kernel type here, elaborating in the docstring.

File: pricers.py [hunk 17]
```diff
@@ -0,0 +1,373 @@
...
+                                                    activation=self.activation)
+            else:                                 
+                self.no_net_target = DeepKernelONetwithPI(branch_layer=self.branch_layers, 
+                                                    trunk_layer=self.trunk_layers, 
+                                                    pi_layer=self.pi_layers, 
+                                                    num_assets=self.dim, 
+                                                    dense=False, 
+                                                    num_outputs=6,
+                                                    activation=self.activation,
+                                                    filters=self.filters,
```
Comment (owenyoung75) [inline pricers.py]: I'm confused: the only difference for dense and non-dense case is this filters and strides clarification?

File: pricers.py [hunk 18]
```diff
@@ -0,0 +1,373 @@
...
+        terminal payoff:
+        in single model case, this is just the payoff terminal function;
+        in multi model cases, this is max(payoff, no_net_target(t,x,u))
+        t, x, u are the ones for the given time interval; 
+        In European case: the corresponded interval is the whole time horizon
+        """
+        payoff = self.option.payoff(t, x, u_hat)
+        return payoff
+
+class EuropeanSolver(MarkovianPricer):
```
Comment (owenyoung75) [inline pricers.py]: Name change to pricer.

File: pricers.py [hunk 19]
```diff
@@ -0,0 +1,373 @@
...
+        if type(self.no_net_target) == DeepONet:
+            y = self.no_net_target((t, x, u))
+        else:
+            u_c, u_p = self.sde.split_uhat(u)
+            y = self.no_net_target((t, x, u_c, u_p))
+        return y
+    
+    def payoff_func(self, t: tf.Tensor, x: tf.Tensor, u_hat: tf.Tensor) -> tf.Tensor:
+        """
+        terminal payoff of each subproblem and when the traiing round is not zero,
```
Comment (owenyoung75) [inline pricers.py]: typo. also, it's better to rephrase as we mentioned earlier: instead of training round, something like exercise_period is more clear.

File: pricers.py [hunk 20]
```diff
@@ -0,0 +1,373 @@
...
+        if self.num_of_time_intervals_for_early_exercise != 0:
+            cont_value = tf.squeeze(self.net_target_forward((t_last, x_last, u_last)), axis=2)
+            early_payoff = self.option.payoff(x, u_hat)
+            payoff = tf.maximum(early_payoff, cont_value)
+        else:
+            payoff = self.option.payoff(x, u_hat)
+            cont = self.net_target_forward((t, x, u_hat))
+        return payoff # (B, M, 1)
+    
+class FixIncomeEuropeanSolver(EuropeanSolver):
```
Comment (owenyoung75) [inline pricers.py]: name change remained...

File: pricers.py
Comment (owenyoung75) [inline pricers.py:1]: Lacking a EarlyExercisePricer here, as we discussed we should move it from the trainer file.

Comment (owenyoung75) [reply pricers.py:1]: Ah ok you still wrote it in a separate file?

File: earlyexercise_pricers.py [hunk 21]
```diff
@@ -0,0 +1,134 @@
+import tensorflow as tf
+from pricers import EuropeanSolver, FixIncomeEuropeanSolver
+from function_space import DeepONet, DeepKernelONetwithPI, DeepKernelONetwithoutPI
+from options import BaseOption
+from sde import ItoProcessDriver, HullWhiteModel
+
+class EarlyExercisePricer:
```
Comment (owenyoung75) [inline earlyexercise_pricers.py]: I think this should be a Pricer class as well, but rather than inheriting the BaseBSDEPrice, which is targeted on single period pricing, this could be defined alone while inherit keras.Model -- good thing is that we could define call(), pay_off() customized fit() etc.

File: earlyexercise_pricers.py [hunk 22]
```diff
@@ -0,0 +1,134 @@
...
+                    t_slice = t[:, :, idx_now:idx_fut+1, :]
+                    x_slice = x[:, :, idx_now:idx_fut+1, :]
+                    dw_slice = dw[:, :, idx_now:idx_fut, :]
+                    u_slice = u[:, :, idx_now:idx_fut+1, :]
+                    return t_slice, x_slice, dw_slice, u_slice
+            sub_dataset = dataset.map(slice_fn)
+        return sub_dataset
+
+
+    def train(self, data: tf.data.Dataset, epochs: int, checkpoint_path: str):
```
Comment (owenyoung75) [inline earlyexercise_pricers.py]: We could just call it fit(), and keep the code same, which would provide more consistency with other pricers.

File: earlyexercise_pricers.py [hunk 23]
```diff
@@ -0,0 +1,134 @@
...
+        """
+        The total training pipeline and we finally attain the no_nets in each sub-time interval.
+        """
+        # learning_rate = self.net_config.lr
+        lr_schedule = tf.keras.optimizers.schedules.ExponentialDecay(
```
Comment (owenyoung75) [inline earlyexercise_pricers.py]: These lines defining scheduler and optimizer and compile, could be put into __init__ instead -- it's a good habit to compile at the beginning to prevent errors.

**Review by owenyoung75 — CHANGES_REQUESTED** [review]

File: pricers.py [hunk 24]
```diff
@@ -0,0 +1,384 @@
...
+        steps = self.eqn_config.time_steps
+        loss_interior = 0.0
+        u_before = u_hat[:, :, :-1, :]
+        x_before = x[:, :, :-1, :]
+        t_before = t[:, :, :-1, :]
+        f = self((t, x, u_hat))
+        grad = self.get_gradient((t, x, u_hat))
+        f_before = f[:, :, :-1, :] # f_before is the value tensor corresponded from time steps 0 to N-1
+        f_after = f[:, :, 1:, :] # f_after is the value tensor corresponded from time steps 1 to N
+        grad = grad[:,:,:-1,:]
```
Comment (owenyoung75) [inline pricers.py]: name grad would be confusing to people in ML. Let's change it to nabla_x.

File: pricers.py [hunk 25]
```diff
@@ -0,0 +1,384 @@
...
+        return self.sde.driver_bsde(t, x, y, u_hat) 
+
+    def diffusion_bsde(self, t: tf.Tensor, x: tf.Tensor, grad: tf.Tensor, dw: tf.Tensor, u_hat: tf.Tensor) -> tf.Tensor:
+        """
+        grad: (B, M, N-1, d)
+        dw: (B, M, N-1, d)
+        give: \sigma(t, x) * grad
+        for a batch of (t, x, par)
+        """
+        if not isinstance(self.sde, HestonModel): # for Heston model, the state is [S, V] so we should truncate the state to just asset
```
Comment (owenyoung75) [inline pricers.py]: this condition is NOT isinstance(), but the comment explains the case isinstance() -- what exactly this condition is trying to achieve?

File: pricers.py [hunk 26]
```diff
@@ -0,0 +1,384 @@
...
+        """
+        t, x, u_hat = inputs
+        if type(self.no_net) == DeepONet:
+            y = self.no_net((t, x, u_hat))
+        else:
+            u_c, u_p = self.sde.split_uhat(u_hat)
+            y = self.no_net((t, x, u_c, u_p))
+        return y
+    
+    def get_gradient(self, inputs: Tuple[tf.Tensor]) -> tf.Tensor:
```
Comment (owenyoung75) [inline pricers.py:175]: Same here, let's change the name to nabla_x rather than gradient.

File: pricers.py (hunk 14 above)
Comment (owenyoung75) [reply pricers.py]: @gengem777 This is an important change. Let's make it a separate function, and also rewrite the loop here -- we could use a tensor-wise operation to replace it (if memory not exploding).

File: pricers.py [hunk 27]
```diff
@@ -0,0 +1,384 @@
...
+    external classes: the sde class which yield the data of input parameters. the option class which yields the
+    input function data and give the payoff function and exact option price.
+    """
+
+    def __init__(self, sde, option, config):
+        super(MarkovianPricer, self).__init__(sde, option, config)
+
+    def loss_fn(self, data: Tuple[tf.Tensor], training=None) -> Tuple[tf.Tensor]:
+        r"""
+        In this method, f is the no_net output value of t, x, u on the time horizon from 0 to T.
```
Comment (owenyoung75) [inline pricers.py:245]: Great func string!!

File: pricers.py (hunk 14 above)
Comment (owenyoung75) [reply pricers.py]: Also, here operations are intense with high-dim tensors, so lets add shape info for each single line in this function -- later when we debug this would be extremely useful.

**Review by owenyoung75 — CHANGES_REQUESTED** [review]
